### ISSUE TITLE
How to log in using an existing JWT token?

### ISSUE THREAD

**Issue by benhubert** [simov/grant#240]
Hi! I have a react application, which uses [react-keycloak](https://www.npmjs.com/package/@react-keycloak/web) and [keycloak-js](https://www.npmjs.com/package/keycloak-js) to log in using keycloak.

From this application, I want to access [directus](https://directus.io), which uses _grant_. This directus instance is configured using the same keycloak instance, as the react application above (but a different client configuration). I can login to directus admin, using keycloak.

_Keycloak_, _directus_ and my _react application_ run under three different subdomains.

Now, my question is, **how do I get an access token for directus, which I can pass with every request**?
Neither the [directus documentation](https://docs.directus.io/reference/api/system/authentication/), nor the grant readme, help me much, at the moment.

What I found out so far:

 * With _react-keycloak_ I already have an access token. But passing this access token with a request to _directus_, always results in HTTP 401 Unauthorized. So, it seems, _directus_ only accepts JWT tokens, generated by _directus_ itself.
 * As far as I understand, I have to call `https://<directus>/auth/oauth/keycloak` to login. But this redirects to my keycloak login, and since I do this call using `fetch`, this ends up in CORS errors for multiple reasons (one of them being the client configuration for _directus_ set to `confidential` instead of `public`).

But why should we do an additional roundtrip to keycloak? Shouldn't it be possible to pass the JWT token of my existing login directly to directus, which then validates this token (optionally using a call to keycloak) and logs me in?

I know, this topic is more about the usage of _grant_ within _directus_ and not about _grant_ itself. But the directus community seems to be a bit too far away from oauth topics, so I hope to find help here now. Maybe the usage of _grant_ adds a solution for my problem to _directus_ out of the box?

**Comment by simov** [comment]
Hi @benhubert, thank you for the thorough explanation.

Lets have a look at a few definitions first:

- Keycloak is your identity provider
- Grant is your OAuth client, that also can act as a proxy
- Directus is your API server, that also hosts Grant

To resolve this issue, you have to use the `https://<directus>/auth/oauth/keycloak` path to login from your React app, and redirect back to it. This will give you the access token that Directus expects.

Let me know if that makes sense.

**Comment by benhubert** [comment]
Thanks for your reply! You mean I should redirect the browser, for example to `https://<directus>/auth/oauth/keycloak?redirect=http://localhost:8000`? With this ...
 * The user gets redirected to keycloak :heavy_check_mark:
 * The user gets redirected to `https://<directus>/auth/oauth/keycloak/callback?session_state...` :heavy_check_mark: 
 * This sets a cookie `directus_refresh_token`, which looks like I am authenticated now. :heavy_check_mark: This cookie is `HttpOnly` and `SameSite: Lax`.

But it seems, this does not help me much. When I now try to fetch 
```
fetch('https://<directus>/items/foo', {
    method: 'GET',
    credentials: 'include',
    cache: 'no-cache',
  }).then(() => {
    ...
  })
```
... I get a `HTTP 403 Forbidden`. As far as I know, I should send an `Authorization: Bearer ...` header with the request (or an `?access_token` request parameter). But the problem is, that I don't have access to this `access_token`.

The documentation says, I could get such a token using the `/auth/refresh` endpoint. But the problem is, that I also don't have access to this `refresh_token`. And the browser does not send cookies with any of these CORS requests.

**Comment by simov** [comment]
The Directus' documentation is a bit scarce to say the least. It's not clear to me what exactly is being returned by the `GET /auth/oauth/:provider` endpoint.

Can you paste here what keys exactly are being returned here by the above endpoint?

**Comment by benhubert** [comment]
It does not "return" directly. It redirects to the oauth provider and back:

```
GET https://<directus>/auth/oauth/keycloak?redirect=http://localhost:8000
HTTP 302 Found
Location: https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/auth?client_id=<directus-client-id>&response_type=code&redirect_uri=https%3A%2F%2F<directus>%2F%2Fauth%2Foauth%2Fkeycloak%2Fcallback
Set-Cookie: connect.sid=s%3A...; Path=/; HttpOnly
```

```
GET https://<keycloak>/auth/realms/<realm>/protocol/openid-connect/auth?client_id=<directus-client-id>&response_type=code&redirect_uri=https%3A%2F%2F<directus>%2F%2Fauth%2Foauth%2Fkeycloak%2Fcallback
HTTP 302 Found
Location: https://<directus>/auth/oauth/keycloak/callback?session_state=396...&code=ffc...
Set-Cookie: [several keycloak cookies]
Set-Cookie: [several keycloak cookies]
```

```
GET https://<directus>/auth/oauth/keycloak/callback?session_state=396...&code=ffc...
HTTP 302 Found
Location: http://localhost:8000
Set-Cookie: directus_refresh_token=xCA...; Max-Age=604800; Path=/; Expires=Mon, 19 Apr 2021 21:41:18 GMT; HttpOnly; SameSite=Lax
```

This is the behavior when doing a _navigation_ on the browser, i.e. entering the first URL into the browser address bar.

If I call it with `fetch`, I get redirected to `https://<keycloak>/auth...` but this returns just HTTP 200 without any further redirect.

**Comment by simov** [comment]
You need to navigate to that URL in the browser. Doing it with fetch won't work. You need to navigate to the Directus login URL from your React app, instead of using the other login that uses another OAuth app. This will give you the access token that Directus expects. You can't do this request in the background using fetch.

**Comment by benhubert** [comment]
Well, that is what I explained above, that it only works with navigating there. But even then ...

> This will give you the access token that Directus expects.

How does it give me the access token? This is exactly the point of my question: How can I get it?

Because, I don't see any possibility, to read this access token from the react application. But without this token, the API is pretty useless for me.

**Comment by simov** [comment]
In this case you need to configure your Grant instance inside Directus with:

```json
{
  "defaults": {
    "callback": "https:// the absolute path to your React app",
    "transport": "querystring"
  }
}
```

That way you will receive the response data as:

```
https://react.app?access_token=...
```

You can configure the `callback` and the `transport` for specific provider/override only or set it dynamically as well, it's not a requirement to set those values inside the `defaults` key.

You can also use the `response` option to trim the response data if needed. All of that can be found in the docs.

Let me know if it works for you.

**Comment by benhubert** [comment]
Many thanks for your help. In the meantime I went for a different solution: I created an API gateway in between, which validates the JWT token, and passes the request to directus. On the directus side I use a single user for all requests with a static API key, which I personally don't like too much, but at least it works. With this solution I also avoid navigating the user to the login, and can do everything directly via `fetch`.

Nevertheless, I really appreciate all your quick responses and detailed explanation. All in all, you helped me a lot with my decision.

I'll close this for now. Thank you!
